openshift/odo
# File sync for devfiles

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:

/kind feature


**What does does this PR do / why we need it**:
This PR implements file syncing for devfile components, using the kubernetes client library. The push code is based on the existing push/sync code used in classic odo, with only a few small tweaks required.

Overall this PR does the following:
- It updates the Devfile deployment code to create a shared emptyDir volume (called `odo-projects`) and mount it under `/projects` in all of the devfile components that set `mountSources: true`
    - This is in line with what Che does for mounting project source
- Creates a `Push()` function in the devfile adapters to handle file syncing
    - If syncing to the pod for the first time (or if `--force` is set), it will sync all of files in the source directory using the `sync` package
    - If syncing to an already running component:
       - it syncs only the files that changed (by checking odo-file-index.json)
       - If any files were deleted from the source directory, it deletes from the remote pod.
- Creates a `DoesComponentExist()` function in the devfile adapters
    - Returns true if a component with the specified label already exists
    - Wrapper for `utils.ComponentExists`
- Moved `IsEmpty` into the `pkg/utils` package from `pkg/component`
    - Needed for both odo "classic" and "devfile" components, so needed to move it to a neutral location.

- Tests
    - Unit tests for new functions
         - Can't really unit test Push(), as it can't be mocked with the fake Kube client (e.g. requires exec'ing to a running pod)
    - Integration tests to cover push and sync scenarios

**Which issue(s) this PR fixes**:

Fixes #2473, https://github.com/openshift/odo/issues/2673

**How to test changes / Special notes to the reviewer**:
1. Build my branch and run `make`
2. Run `make test` to verify unit tests pass
3. Enable experimental mode
4. Run `odo push` verify that the deployment starts and files get synced
5. Change/add/remove files, verify the proper changes get reflected on the container.
